### PR TITLE
[FEATURE] Improve robustness of differentiable rigid body simulation.

### LINKED ISSUE #2069: [Bug]: No gradients wrt control force
### Bug Description

I am trying to get gradients wrt control inputs, but the gradient graph seems to be broken. I can get grads wrt initial pose (as in the unit test) but not for control.

### Steps to Reproduce

```
import time
import numpy as np
import torch
import genesis as gs
from etils import epath
from torchviz import make_dot

gs.init(logging_level="warning", backend=gs.cuda)

dt = 1e-2
horizon = 10
substeps = 1
num_iter = 400
lr = 5e-2

scene = gs.Scene(
    sim_options=gs.options.SimOptions(
        dt=dt,
        substeps=substeps,
        requires_grad=True,
        gravity=(0.0, 0.0, -9.81),
    ),
    rigid_options=gs.options.RigidOptions(
        enable_collision=False,
        enable_self_collision=False,
        enable_joint_limit=False,
        disable_constraint=False,
        use_contact_island=False,
        use_hibernation=False,
    ),
    viewer_options=gs.options.ViewerOptions(
        camera_pos=(0.0, -4.0, 1.5),
        camera_lookat=(0.0, 0.0, 1.0),
        camera_fov=30.0,
        res=(960, 640),
        max_FPS=240,
    ),
    show_viewer=True,
)

mjcf_path = str(epath.resource_path('genesis_playground') / "envs/xmls/cartpole.xml")

cartpole = scene.add_entity(
    gs.morphs.MJCF(file=mjcf_path),
)

# Joints: slider (cart x), hinge_1 (pole angle)
slider_joint = cartpole.get_joint("slider")
hinge_joint = cartpole.get_joint("hinge_1")

# Local q indices (for generalized coordinates)
slider_q_idx = slider_joint.q_idx_local
hinge_q_idx = hinge_joint.q_idx_local

scene.build()
cartpole.set_dofs_armature([0.0, 0.0], [slider_q_idx, hinge_q_idx])
cartpole.set_dofs_force_range(
    lower=np.array([-10.0, -10.0]),
    upper=np.array([10.0, 10.0]),
    dofs_idx_local=[slider_q_idx, hinge_q_idx]
)


# Learnable initial slider position and hinge angle
q_init = gs.tensor([1.0, 0.5], requires_grad=True)
q = q_init.detach()
q.requires_grad_()
control = gs.tensor([3.0], requires_grad=True)
optimizer = torch.optim.Adam([control, q], lr=lr)
scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(
    optimizer, T_max=num_iter, eta_min=1e-3
)

# Optimization loop
for it in range(num_iter):
    # scene.reset()
    cartpole.set_qpos(q)
    cartpole.set_dofs_velocity(control, slider_q_idx)
    
    # Rollout
    for i in range(horizon):
        scene.step()

    # Final state
    q_final = cartpole.get_qpos()
    x_T = q_final[slider_q_idx]
    theta_T = q_final[hinge_q_idx]

    # Loss: ONLY make pole upright at final time (theta_T -> 0)
    loss = theta_T**2 + x_T**2
    optimizer.zero_grad()


    make_dot(loss, 
             params={
                 "q": q, 
                 "control": control, 
                 "q_final": q_final, 
                 "x_T": x_T, 
                 "theta_T": theta_T,
                 "loss": loss,
                 },
             show_attrs=True,
             show_saved=True).render("autograd_graph", format="png")


    # retain_graph=True so Genesis' internal graph is not freed
    loss.backward(retain_graph=True)
    optimizer.step()
    scheduler.step()


    print(
        f"[iter {it:03d}] "
        f"grad={q.grad}, "
        f"ctrl grad={control.grad}, "
    )
```

### Expected Behavior

Gradients not none

### Screenshots/Videos

Image

### Relevant log output

```shell
[iter 000] grad=tensor([2.0284, 1.0394], device='cuda:0'), ctrl grad=None, 
[iter 001] grad=tensor([1.9287, 0.9310], device='cuda:0'), ctrl grad=None, 
[iter 002] grad=tensor([1.8292, 0.8229], device='cuda:0'), ctrl grad=None, 
[iter 003] grad=tensor([1.7300, 0.7156], device='cuda:0'), ctrl grad=None, 
[iter 004] grad=tensor([1.6313, 0.6097], device='cuda:0'), ctrl grad=None,
```

### Environment

- OS: [e.g. Ubuntu 24.04, Windows 11 24H2]
- GPU/CPU [e.g. A100, RTX 4090, M3pr, Intel I9-9900k, Ryzen 5900x] (N/A if no GPU/CPU)
- GPU-driver version (N/A if no GPU)
- CUDA / CUDA-toolkit version (N/A if non-Nvidia)

### Release version or Commit ID

a2787f5fa1aeac250c4f36702b5de949993ec057

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by SonSang**
## Description

This PR improves the robustness of the differentiable rigid body simulation, and add in-depth unit tests to guarantee the differentiability in various scenarios. This PR is necessary to solve optimization problems that leverage the differentiability of the rigid body simulation.

## Related Issue

Resolves https://github.com/Genesis-Embodied-AI/genesis-world/issues/2069

## Motivation and Context

Previously, the backward pass of rigid body simulation was brittle, and there was not enough unit tests that verify the backward pass in diverse scenarios. Because of that, several edge cases in the backward pass were not caught and they silently failed downstream applications that leverage the differentiability.

## How Has This Been / Can This Be Tested?

To test the differentiability of the rigid body simulation, expanded the unit tests in the existing `test_grad.py` and split them into the 4 different files as shown below.
```
tests/test_grad_fd.py
tests/test_grad_optim.py
tests/test_grad_utils.py
tests/test_grad_mpm.py
```
Each of these tests cover:
* `test_grad_fd.py`: Contains unit tests for verifying the correctness of the (analytical) gradients from the backward pass by comparing them with the gradients that we estimate from finite difference method. We test under various topology (e.g. cartpole, hopper, etc.) to catch as much edge cases as possible. It tests not only for the single step, but also gradients over multiple steps. Also, it tests under various constraint scenarios, such as joint limit or collision.
* `test_grad_optim.py`: Contains unit tests that optimize a certain loss value by optimizing input values (e.g. velocity, force). These tests serve as the proxy for the downstream optimization tasks based on differentiability.
* `test_grad_utils.py`: Check utility functions that are used for the backward pass. For now, there is a single test that verifies if the `backward` function of the `Scene` works as expected.
* `test_grad_mpm.py`: Unlike other tests that check rigid body simulation, this file contains tests for MPM solver.

## Checklist:

- [x] I read the **CONTRIBUTING** document.
- [x] I followed the `Submitting Code Changes` section of **CONTRIBUTING** document.
- [x] I tagged the title correctly (including BUG FIX/FEATURE/MISC/BREAKING)
- [ ] I updated the [documentation](https://github.com/Genesis-Embodied-AI/genesis-doc) accordingly or no change is needed.
- [x] I tested my changes and added instructions on how to test it for reviewers.

- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by pschoc** [comment]
any updates on this PR? I'm currently looking into diff RL with the Franka Emika Panda, but with the current main branch, the slide joint / prismatic joint is producing NaN gradients. This branch works for me, so I'm looking forward to it being finalized and merged.

**Comment by duburcqa** [comment]
@pschoc There is no plan to merge this PR at the moment. The maintenance burden is too high for the benefits. We need to improve autodiff support on Quadrants side.

**Comment by pschoc** [comment]
@duburcqa alright, thanks for the clarification

**Comment by duburcqa** [comment]
@pschoc Actually I changed my mind. It is not as bad as I thought after cleanup, and coverage is solid, so I will merge after finishing cleaning up.

**Comment by pschoc** [comment]
@duburcqa awesome! Really looking forward 🚀

**Comment by duburcqa** [comment]
@codex review

File: genesis/engine/solvers/rigid/abd/manual_bw.py [hunk 1]
```diff
@@ -0,0 +1,529 @@
...
+                    # Final pose is set absolutely from qpos (slot in unused);
+                    # xanchor = qpos[0:3].
+                    for j in qd.static(range(3)):
+                        rigid_info.qpos.grad[q_start + j, i_b] = (
+                            rigid_info.qpos.grad[q_start + j, i_b] + g_pos[j] + xanchor_grad[j]
+                        )
+                    for j in qd.static(range(4)):
+                        rigid_info.qpos.grad[q_start + 3 + j, i_b] = (
+                            rigid_info.qpos.grad[q_start + 3 + j, i_b] + g_quat[j]
+                        )
```
[the remark this answers is not in the document]
Comment (duburcqa) [reply genesis/engine/solvers/rigid/abd/manual_bw.py:98]: fixed

File: genesis/utils/geom.py [hunk 2]
```diff
@@ -270,6 +320,46 @@ def qd_transform_by_quat(v, quat):
...
+    d_out2_d_quat = 2.0 * qd.Vector(
+        [
+            -q_y * v_x + q_x * v_y + q_w * v_z,
+            q_z * v_x + q_w * v_y - q_x * v_z,
+            -q_w * v_x + q_z * v_y - q_y * v_z,
+            q_x * v_x + q_y * v_y + q_z * v_z,
+        ],
+        dt=gs.qd_float,
+    )
+    return out_grad[0] * d_out0_d_quat + out_grad[1] * d_out1_d_quat + out_grad[2] * d_out2_d_quat
```
[the remark this answers is not in the document]
Comment (duburcqa) [reply genesis/utils/geom.py]: fixed
